docs/proposals/protobuf.md

```diff
@@ -0,0 +1,501 @@
+}
+
+// Affinity is a group of affinity scheduling rules, currently
+// only node affinity, but in the future also inter-pod affinity.
+message Affinity {
+  // Describes node affinity scheduling rules for the pod.
+  optional NodeAffinity nodeAffinity = 1 [(gogoproto.customname) = "NodeAffinity"];
+}
+```
+
+## Wire format
+
+In order to make Protobuf serialized objects recognizable in a binary form,
+the encoded object must be prefixed by a magic number, and then wrap the
+non-self-describing Protobuf object in a Protobuf object that contains
+schema information.  The protobuf object is referred to as the `raw` object
+and the encapsulation is referred to as `wrapper` object.
+
+The simplest serialization is the raw Protobuf object with no identifying
+information. In some use cases, we may wish to have the server identify the
+raw object type on the wire using a protocol dependent format (gRPC uses
+a type HTTP header). This works when all objects are of the same type, but
+we occasionally have reasons to encode different object types in the same
+context (watches, lists of objects on disk, and API calls that may return
+errors).
+
+To identify the type of a wrapped Protobuf object, we wrap it in a message
+in package `k8s.io/kubernetes/pkg/runtime` with message name `Unknown`
+having the following schema:
+
+    message Unknown {
+      optional TypeMeta typeMeta = 1;
+      optional bytes value = 2;
+      optional string contentEncoding = 3;
+      optional string contentType = 4;
+    }
+
+    message TypeMeta {
+      optional string apiVersion = 1;
+      optional string kind = 2;
+    }
+
+The `value` field is an encoded protobuf object that matches the schema
+defined in `typeMeta` and has optional `contentType` and `contentEncoding`
+fields.  `contentType` and `contentEncoding` have the same meaning as in
+HTTP, if unspecified `contentType` means "raw protobuf object", and
+`contentEncoding` defaults to no encoding. If `contentEncoding` is
+specified, the defined transformation should be applied to `value` before
+attempting to decode the value. Some objects may in the future encode
+JSON into the value field (for example, templates or third party objects).
+
+A client should use the `kind` and `apiVersion` fields to identify the
+correct protobuf IDL for that message and version, and then decode the
+`bytes` field into that Protobuf message.
+
+Any Unknown value written to stable storage will be given a 4 byte prefix
+`0x6b, 0x38, 0x73, 0x00`, which correspond to `k8s` followed by a zero byte.
+The content-type `application/vnd.kubernetes.protobuf` is defined as
+representing the following schema:
+
+    MESSAGE = '0x6b 0x38 0x73 0x00' UNKNOWN
+    UNKNOWN = <protobuf serialization of k8s.io/kubernetes/pkg/runtime#Unknown>
+
+A client should check for the first four bytes, then perform a protobuf
+deserialization of the remaining bytes into the `runtime.Unknown` type.
```

Comment:
Yeah, pkg/runtime/serializer/recognizer (used by the UniversalDeserializer)
will use the prefix bytes to peek at the head of the object and make a
decision whether to try a decode.

On Mon, Mar 7, 2016 at 4:11 AM, Wojciech Tyczynski <notifications@github.com
> wrote:

> In docs/proposals/protobuf.md
> <https://github.com/kubernetes/kubernetes/pull/22600#discussion_r55183244>
> :
>
> > +JSON into the value field (for example, templates or third party objects).
> > +
> > +A client should use the `kind` and `apiVersion` fields to identify the
> > +correct protobuf IDL for that message and version, and then decode the
> > +`bytes` field into that Protobuf message.
> > +
> > +Any Unknown value written to stable storage will be given a 4 byte prefix
> > +`0x6b, 0x38, 0x73, 0x00`, which correspond to `k8s` followed by a zero byte.
> > +The content-type `application/vnd.kubernetes.protobuf` is defined as
> > +representing the following schema:
> > +
> > +    MESSAGE = '0x6b 0x38 0x73 0x00' UNKNOWN
> > +    UNKNOWN = <protobuf serialization of k8s.io/kubernetes/pkg/runtime#Unknown>
> > +
> > +A client should check for the first four bytes, then perform a protobuf
> > +deserialization of the remaining bytes into the `runtime.Unknown` type.
>
> And if the first four bytes are different, then we will continue decoding
> to JSON, right?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/22600/files#r55183244>.
>